### PR TITLE
[CSM] Add function to set the FOV of the local player

### PR DESCRIPTION
**Pull request by bigfoot547**

### CODE REVIEW

**Comment by nerzhul** [comment]
Thanks but you missed the feature merge window, it will be delayed on next release

**Comment by bigfoot547** [comment]
I understand. :'(

File: doc/client_lua_api.md [hunk 1]
```diff
@@ -721,12 +721,6 @@ Call these functions only at load time!
 ### Player
 * `minetest.get_wielded_item()`
     * Returns the itemstack the local player is holding
-* `minetest.send_chat_message(message)`
-    * Act as if `message` was typed by the player into the terminal.
-* `minetest.run_server_chatcommand(cmd, param)`
-    * Alias for `minetest.send_chat_message("/" .. cmd .. " " .. param)`
-* `minetest.clear_out_chat_queue()`
```
Comment (nerzhul) [inline doc/client_lua_api.md]: why remove this ?

File: src/camera.h [hunk 2]
```diff
@@ -91,6 +92,36 @@ class Camera
...
+	{
+		m_cache_zoom_fov = fov;
+	}
+	
+	inline f32 getFov()
+	{
+		return m_cache_fov;
+	}
+	
+	inline f32 getZoomFov()
```
Comment (nerzhul) [inline src/camera.h]: all getter should be const functions

File: doc/client_lua_api.md (hunk 1 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: @nerzhul It was a mistake when copying stuff... I will fix it in a future commit.

**Comment by SmallJoker** [comment]
What about the zoom feature we've already got? This PR in combination with a key press getter function will make the required privilege superflous.

**Comment by DS-Minetest** [comment]
@SmallJoker The minimum value of zoom fov is much lower than the one of the normal fov.
Or aren't the added functions limited? That would be great IMO!

**Comment by nerzhul** [comment]
Okay for me

File: src/camera.h [hunk 3]
```diff
@@ -91,6 +92,36 @@ class Camera
 		return m_camera_offset;
 	}
 
+	inline void setFov(f32 fov)
+	{
+		m_cache_fov = fov;
+	}
```
Comment (SmallJoker) [inline src/camera.h]: Quite nonsense to have getter and setter functions for `m_cache_fov` and `m_cache_zoom_fov`. If there were boundary checks, this would be acceptable but how it is right now you could make them public as `cache_zoom`.

**Comment by nerzhul** [comment]
Please let members private it's an object not a structure.

**Comment by bigfoot547** [comment]
@SmallJoker the range *is* limited because of [this, right here](https://github.com/bigfoot547/minetest/blob/670abb8c577f730bc57b7fb635eff145a5e5cb3b/src/camera.cpp#L393-L400).

**Comment by SmallJoker** [comment]
@nerzhul Well, we already have public members for objects. The best example for this is the `Player` class.

@bigfoot547 I'm talking about limits when setting the `cache_fov`/`cache_zoom_fov` variables to a new value. The range limit is done in each `Camera::update` call and does not affect the saved settings at all. That's why I would prefer to expose this variable as `public` instead of adding functions.

**Comment by nerzhul** [comment]
@SmallJoker yes i know we have that ugly interface in player; i started to cleaning up years ago by moving some attributes to PlayerSAO/RemotePlayer or LocalPlayer for client, but not finished yet. Having getter and setters has no performance impact on code (they are inlined in majority cases), and we respect properly the difference between a struct and a class (class intend to have private members and struct public members)

File: doc/client_lua_api.md [hunk 4]
```diff
@@ -842,16 +842,25 @@ Please do not try to access the reference until the camera is initialized, other
     * Returns with same syntax as above
 * `get_fov()`
     * Returns:
-
+
```
Comment (nerzhul) [inline doc/client_lua_api.md]: can you fix those spaces ?

File: doc/client_lua_api.md [hunk 5]
```diff
@@ -842,16 +842,25 @@ Please do not try to access the reference until the camera is initialized, other
...
 ```
 
+* `set_fov(fov_degrees)`
+	* Sets the FOV for the local player
+* `set_zoom_fov(fov_degrees)`
+	* Sets the FOV for the local player when zoomed in
+* `save_fov()`
+	* Sets the "fov" setting to `get_fov().actual`
+* `save_zoom_fov()`
+	* Sets the "zoom_fov" setting to `get_fov().zoom`
```
Comment (ShadowNinja) [inline doc/client_lua_api.md]: Just add a `persist` boolean argument to the FOV setters.

Comment (paramat) [inline doc/client_lua_api.md]: Why not just make 'set' set the zoom settings permanently as with any other setting?
Why have the complexity of a temporary plus a way to save it?

**Comment by paramat** [comment]
A description in the first post would be nice, explaining basic implementation, otherwise it's difficult to review.

File: doc/client_lua_api.md (hunk 5 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: Done

**Comment by paramat** [comment]
Since this is CSM, altering FOV must surely need the 'zoom' priv? The current zoom is clientside which is why a priv was added for that.
This is also an advantage that needs to be restricted by flavours.

**Comment by red-001** [comment]
why do we even have a zoom privilege in the first place?

**Comment by paramat** [comment]
See the PR thread that added zoom. Zoom adds a 'superpower' to a player so a restriction was required for survival such that a player would need to craft a telescope type item before being able to zoom.

**Comment by octacian** [comment]
I personally don't think this is an issue. For that matter, changing FOV in the advanced settings menu should require the zoom privilege, which would both be hard to code and wouldn't make sense whatsoever. I would personally suggest that the `zoom` privilege be removed and instead a `set_fov` (and `get_fov`) method be added to the PlayerSAO server-side so as to allow the server to force a specific FOV at any point in time.

The functionality introduced by this PR could then either be controlled by the flavor PR or maybe a physics override.

**Comment by paramat** [comment]
> changing FOV in the advanced settings menu should require the zoom privilege

Good point, but doing so is a permanent change not very useful for gameplay, to gain an advantage normal FOV could indeed be set small, but is not useful if continuous. An advantage only happens when you can switch from normal to 'fov_zoom' in-game, and that requires a priviledge.

**Comment by bigfoot547** [comment]
Fixed compile and LINT

**Comment by paramat** [comment]
We are reconsidering the zoom priv and whether to replace with a HUD flag.
How should this feature be controlled, with the zoom priv or by CSM flavours?
bigfoot547 i guess from our CSM flavours discussion you don't want any restriction?

**Comment by red-001** [comment]
just limit it to the same range the fov setting is limited to.

**Comment by paramat** [comment]
That isn't a restriction, that still allows a zoom as powerful as 'Z'.
We already restrict zoom with a priv, so this also needs restricting.

File: src/script/lua_api/l_camera.cpp [hunk 6]
```diff
@@ -64,13 +64,62 @@ int LuaCamera::l_get_fov(lua_State *L)
...
+	if (!camera)
+		return 0;
+
+	if (!lua_isnumber(L, 2))
+		return 0;
+
+	if (!lua_isboolean(L, 3))
+		persist = false;
+	else
+		persist = lua_toboolean(L, 3);
```
Comment (red-001) [inline src/script/lua_api/l_camera.cpp]: just replace this whole segment with `bool persist = lua_toboolean(L, 3);`

File: src/script/lua_api/l_camera.cpp [hunk 7]
```diff
@@ -64,13 +64,62 @@ int LuaCamera::l_get_fov(lua_State *L)
...
+
+	fov = lua_tonumber(L, 2);
```
Comment (red-001) [inline src/script/lua_api/l_camera.cpp]: `float fov = luaL_checknumber(L, 2)`

File: src/script/lua_api/l_camera.cpp [hunk 8]
```diff
@@ -64,13 +64,62 @@ int LuaCamera::l_get_fov(lua_State *L)
...
+{
+	Camera *camera = getobject(L, 1);
+
+	bool persist;
+	float fov;
+
+	if (!camera)
+		return 0;
+
+	if (!lua_isnumber(L, 2))
```
Comment (red-001) [inline src/script/lua_api/l_camera.cpp]: get rid of this.

File: src/script/lua_api/l_camera.cpp [hunk 9]
```diff
@@ -64,13 +64,62 @@ int LuaCamera::l_get_fov(lua_State *L)
...
+		return 0;
+
+	if (!lua_isboolean(L, 3))
+		persist = false;
+	else
+		persist = lua_toboolean(L, 3);
+
+	fov = lua_tonumber(L, 2);
+	camera->setZoomFov(fov, persist);
```
Comment (red-001) [inline src/script/lua_api/l_camera.cpp]: same as above

**Comment by paramat** [comment]
Latest change restricts the normal FOV to a minimum of 30, so that the zoom priviledge is still needed to gain any advantage, thanks and good idea, removing my -1. This will therefore need no additional restriction.

**Comment by bigfoot547** [comment]
@paramat @SmallJoker Can someone please review this? Thanks!

**Comment by paramat** [comment]
Sorry i avoid CSM, don't understand it enough.

File: doc/client_lua_api.md [hunk 10]
```diff
@@ -848,10 +848,17 @@ Please do not try to access the reference until the camera is initialized, other
          x = number,
          y = number,
          max = number,
-         actual = number
+         actual = number,
```
Comment (red-001) [inline doc/client_lua_api.md]: You should explain these better.

File: doc/client_lua_api.md [hunk 11]
```diff
@@ -848,10 +848,17 @@ Please do not try to access the reference until the camera is initialized, other
...
+         zoom = number
      }
 ```
 
+* `set_fov(fov_degrees, [persist])`
+	* Sets the FOV for the local player
+	* Pass true to the second argument to save the zoom fov to the setting.
+* `set_zoom_fov(fov_degrees [persist])`
```
Comment (SmallJoker) [inline doc/client_lua_api.md]: `,` missing before space

**Comment by rubenwardy** [comment]
What are the usecases for this?

File: doc/client_lua_api.md (hunk 10 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: I'm trying to figure out how to make this table less ambiguous...

**Comment by bigfoot547** [comment]
@rubenwardy What aren't the usecases for this? :laughing:

**Comment by paramat** [comment]
Please could you add a description? Ultra-simple PRs may not need one but this is not and the point about restricting range of FOV in each mode, to avoid zoom-cheating, is important.

I feel it's good that rubenwardy is checking for usecases, this is what we need to do for any new CSM function so that they are added for a reason, and to avoid unnecessary functions that could be harmful to a server and therefore wold need another restriction added.

What are the advantages of clientside FOV change?

File: doc/client_lua_api.md (hunk 11 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: Good catch. I'll fix this along with the previous review.

**Comment by octacian** [comment]
One possible use could be as an effect to apply to the player when sprinting, although the rest of the needed APIs to do such a thing are not yet in place. For example, though, the FOV could be slightly widened when sprinting, and seeing as it's client-side, one could even change the FOV in steps to make it look smooth.

File: doc/client_lua_api.md [hunk 12]
```diff
@@ -845,13 +845,20 @@ Please do not try to access the reference until the camera is initialized, other
...
      {
-         x = number,
-         y = number,
-         max = number,
-         actual = number
+         x = number,       -- Horizontal FOV
+         y = number,       -- Vertical FOV
+         max = number,     -- The maximum of x and y
+         actual = number,  -- What the FOV of the player would be (not adjusted for zooming)
+         zoom = number     -- What the FOV of the player when zooming would be
```
Comment (paramat) [inline doc/client_lua_api.md]: Instead of 'number', 'float' or 'integer' would be better.

Comment (paramat) [inline doc/client_lua_api.md]: And document if FOV is in degrees or radians.

Comment (bigfoot547) [inline doc/client_lua_api.md]: Okay

Comment (bigfoot547) [inline doc/client_lua_api.md]: >_< I just pushed 5 seconds ago.
Should I make yet another commit?
If I do, maybe you should cancel the travis builds.

Comment (bigfoot547) [inline doc/client_lua_api.md]: Done

**Comment by paramat** [comment]
Please could you document whether these are degrees or radians?

**Comment by bigfoot547** [comment]
done.

File: doc/client_lua_api.md [hunk 13]
```diff
@@ -841,17 +841,25 @@ Please do not try to access the reference until the camera is initialized, other
...
 
 ```lua
      {
-         x = number,
-         y = number,
-         max = number,
-         actual = number
+         x = float,        -- Horizontal FOV (This is not a whole number as it is converted from radians)
+         y = float,        -- Vertical FOV (This is not a whole number as it is converted from radians)
+         max = float,      -- The maximum of x and y (This is not a whole number as it is converted from radians)
```
Comment (paramat) [inline doc/client_lua_api.md]: "This is not a whole number" is not necessary because it is obvious from 'float'.
"converted from radians" is not necessary to know, that's a technical point.
Are these degrees or radians? "converted from radians" does not clearly mean 'degrees'.

File: doc/client_lua_api.md [hunk 14]
```diff
@@ -841,17 +841,25 @@ Please do not try to access the reference until the camera is initialized, other
...
+         actual = integer, -- What the FOV of the player would be (not adjusted for zooming)
+         zoom = integer    -- What the FOV of the player when zooming would be
```
Comment (paramat) [inline doc/client_lua_api.md]: Degrees or radians for these 2?

File: doc/client_lua_api.md (hunk 13 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: https://github.com/minetest/minetest/pull/5797/files#diff-927bb329aa13096bcf9123248a422bc2R844

File: doc/client_lua_api.md (hunk 14 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: https://github.com/minetest/minetest/pull/5797/files#diff-927bb329aa13096bcf9123248a422bc2R844

File: doc/client_lua_api.md (hunk 13 above)
Comment (bigfoot547) [inline doc/client_lua_api.md]: I will remove `Horizontal FOV (This is not a whole number as it is converted from radians)` now though

**Comment by paramat** [comment]
All you need is 'in degrees' added to all lines except 'max' where it is obvious from the first 2.

**Comment by bigfoot547** [comment]
What's wrong with the current implementation?

**Comment by paramat** [comment]
Sorry didn't see the line above, seems ok then.

**Comment by bigfoot547** [comment]
Thanks @paramat

File: doc/client_lua_api.md [hunk 15]
```diff
@@ -841,17 +841,25 @@ Please do not try to access the reference until the camera is initialized, other
 * `get_camera_mode()`
     * Returns with same syntax as above
 * `get_fov()`
+	* All values are in degrees
```
Comment (paramat) [inline doc/client_lua_api.md]: Indentation should match other lines in this section.

File: doc/client_lua_api.md [hunk 16]
```diff
@@ -841,17 +841,25 @@ Please do not try to access the reference until the camera is initialized, other
...
+         zoom = integer    -- What the FOV of the player when zooming would be
      }
 ```
 
+* `set_fov(fov_degrees, [persist])`
+	* Sets the FOV for the local player
+	* Pass true to the second argument to save the zoom fov to the setting.
+* `set_zoom_fov(fov_degrees, [persist])`
+	* Sets the FOV for the local player when zoomed in
+	* Pass true to the second argument to save the zoom fov to the setting.
```
Comment (paramat) [inline doc/client_lua_api.md]: Indentation should match other lines in this section.

File: doc/client_lua_api.md [hunk 17]
```diff
@@ -841,17 +841,25 @@ Please do not try to access the reference until the camera is initialized, other
...
      {
-         x = number,
-         y = number,
-         max = number,
-         actual = number
+         x = float,        -- Horizontal FOV
+         y = float,        -- Vertical FOV
+         max = float,      -- The maximum of x and y
+         actual = integer, -- What the FOV of the player would be (not adjusted for zooming)
+         zoom = integer    -- What the FOV of the player when zooming would be
```
Comment (paramat) [inline doc/client_lua_api.md]: What's the meaning of 'would be' in these lines? How is it different from x,y? This needs clarifying.

**Comment by bigfoot547** [comment]
@paramat: I have rebased and addressed your reviews, so can you please review again?

**Comment by bigfoot547** [comment]
needs a rebase again...

**Comment by bigfoot547** [comment]
Rebased. @SmallJoker @paramat Please review. I don't want this PR to sit doing nothing :L

**Comment by rubenwardy** [comment]
should be added to CSM flavours.

**Comment by bigfoot547** [comment]
Ok, I'll look into it

**Comment by paramat** [comment]
:-1: nerzhul this can't be merged because it conflicts with the zoomFOV changes being made by lhofhansl and me, and has errors.
See #6650 , this PR is low priority and should wait until that is merged, and then reworked around the new approach.

In this PR camera FOV is now restricted to 7 to 160 when it needs to range from 1 to 160.
Per-player ZoomFOV has to be overridable by the server when it is set by #6650 
Client-settable ZoomFOV must be limited to a minimum of 15 degrees as in #6650 , because zoom now loads distant terrain which puts a load on the server and gives the player an advantage of being able to generate and load terrain beyond the server's maximum generate and send ranges.
Client-settable zoom should not affect the server and cause it to generate or load terrain beyond these ranges.

Is this useful for server-provided CSM? That is our new decider on CSM features, as CSM has to be as minimal as possible. If it is, it needs to be controlled by flavors as it is something that gives a player an advantage: if the client has distant mapblocks loaded, it can look at those and magnify them.

:-1: for now.

**Comment by paramat** [comment]
:-1: Actually, sorry but this can't be merged in any form, because zoom now loads and/or generates distant world like a telescope, a powerful ability that must be under complete server control, it can't be added to CSM.

**Comment by bigfoot547** [comment]
:/ well, I guess there's only one thing to say.
https://youtube.com/watch?v=cdEQmpVIE4A
